### ISSUE TITLE
New example - Top Down Shooter Template

### ISSUE THREAD

**Issue by UlisesFreitas** [GDevelopApp/GDevelop-examples#15]
## Describe the example
![ITop Down Shooter](https://www.gamesigniter.com/topdownshooter/screenshot-1.png)

Top Down Shooter Template for GDevelop 5
A simple template created for GDevelop 5 to demonstrate how you can make a Top Down shooter.
Controls:
Use the mouse to aim
A - Left
D - Right
W - Up
S - Down

## Game folder
https://www.gamesigniter.com/topdownshooter/top-down-shooter-source-code.zip
or here the source code
https://github.com/UlisesFreitas/top-down-shooter-template

## Screenshots
https://www.gamesigniter.com/topdownshooter/desktop-icon-512.png
https://www.gamesigniter.com/topdownshooter/screenshot-1.png
https://www.gamesigniter.com/topdownshooter/screenshot-2.png
https://www.gamesigniter.com/topdownshooter/screenshot-3.png

**Comment by Midhil457** [comment]
Do you have the rights to use the assets? Are the assets used proprietary? What is the licence?

- There is an aseprite files in the game folder. Please remove them
- To get a thumbnail in the example's library, you have to add a image called `thumbnail.png` and not the icons for platforms (But, these are welcome)
- Please remove the screenshots included in the game folder
- There are many assets that are not used but included in the folder. Please remove those
- Add a proper package name in the properites panel. com.gdevelop.shooty would be fine
- You have included extensions that are not used. Please remove them

- There are some z-order issues in many scenes (Ex: the cursour is below some objects)
- I don't think the prefix **obj** is required for all the objects
- Same for the **gs** prefix for variables
- Add proper comments that explain the events
- set the orgin point of the cursour to center. Currently, the bullets are being shot to the corner of the cursour
- Make the mountain a panel sprite. This will help reduce the sprites into one single object
- The cursour doesn't work in the died screen
- Why does the enemy have a platformer behavior?
- Try to color the event groups better
- the enemy AI seems bad. Could you try to make it more structured. A state machine would work great
- Same for the Game State, use a state machine that handles the pause state, died state, etc... and have it under a single events group
- I would really suggest restructuring the events. I found that many related events are scattered everywhere

I have added comments with my review of the events in the project: 
[Shooty.zip](https://github.com/GDevelopApp/GDevelop-examples/files/6725577/Shooty.zip)

I might have to make multiple reviews as this example needs some changes (not big ones. Many small things)
:)

**Comment by Silver-Streak** [comment]
Hey @Midhil457 

I'm going to take over some of the review asks for Ulises. Wanted to clarify some items before I started on it sometime later this week:

> * I don't think the prefix **obj** is required for all the objects
> * Same for the **gs** prefix for variables

While these aren't required, I'm not seeing how there's any huge negatives to this being in the example, especially since some examples have object names/variables in other languages. Unless you say otherwise, I don't think this needs changing

> * Try to color the event groups better

Gonna need context on this one, the majority of examples in the engine, including all of the template games, just use standard group colors. Is there something you're looking for on this one?


> * the enemy AI seems bad. Could you try to make it more structured. A state machine would work great
> * Same for the Game State, use a state machine that handles the pause state, died state, etc... and have it under a single events group

I'm not 100% sure this is valid feedback for a base game template/example, unless the example is explicitly around state machines, or designed as an "advanced example", state machines seem like a bit more than what is necessary. As an example the retro shooter has everything on one event sheet, not even external events, but is one of the cleanest examples we have today.

Again, I'm not a reviewer, so if you're adamant that these need to be included for this to be approved in the engine, I can take another look.


The rest of the items make sense to me and I'll do what I can.

**Comment by Midhil457** [comment]
> While these aren't required, I'm not seeing how there's any huge negatives to this being in the example, especially since some examples have object names/variables in other languages. Unless you say otherwise, I don't think this needs changing

I suggested removing it as it didn't really helped with anything. All it did was make object names repetitive. It's really obvious to find an object, obj prefix doesn't help in anything. But, I am leaving it to you to decide :)

> I'm not 100% sure this is valid feedback for a base game template/example, unless the example is explicitly around state machines, or designed as an "advanced example", state machines seem like a bit more than what is necessary. As an example the retro shooter has everything on one event sheet, not even external events, but is one of the cleanest examples we have today.

Currently the enemy always follows the player, not in a efficient nor in a reliable way. I still don't understand how the enemies move. But, some enemies get stuck between obstacles. The enemies don't use pathfinding behavior (instead a platformer behavior?!) I recommended a state machine because they work reliably, easy to read and make, and can be modified easily by anyone. As this also serves as a template, something that works well would be a good idea. Something that is basic and just works most of the time would be more than enough

For the Game State, same reasons for the state machine. It should easy and understandable with 3 states, GameActive, Died, Paused. I had a great success with using a state machine in the breakout example. 

>- Try to color the event groups better
Gonna need context on this one, the majority of examples in the engine, including all of the template games, just use standard group colors. Is there something you're looking for on this one?

Coloring comments and groups are fine. What I mean is that it seems that most object group/comments are colored semi-randomly, which makes the event sheet look chaotic (ignore the orange comments in the screenshot, those are my review comments)

![C2](https://user-images.githubusercontent.com/73597906/123780815-369e6400-d8f1-11eb-9b8d-e24442e2d5e5.PNG)
![C1](https://user-images.githubusercontent.com/73597906/123780822-3900be00-d8f1-11eb-972b-e46dfd0fa006.PNG)

Also, sorry if I have gone too far and complaining about small things in the comments I have added in the project :)
:)

**Comment by 4ian** [comment]
Thanks all, if we can streamline/simplify a bit this this will be a great example :)

**Comment by Silver-Streak** [comment]
I haven't forgotten about this. I just haven't had time to dig into it yet. I hope to look through it this week.

**Comment by 4ian** [comment]
That would be really cool :) Thank you!

**Comment by UlisesFreitas** [comment]
Hi mates can anyone tell me what needs to be done here I'm a bit lost of what issues needs. to be fixed.

**Comment by Midhil457** [comment]
I have included a copy of the project with all my suggestions in the first review

**Comment by Silver-Streak** [comment]
@UlisesFreitas This is probably waiting on me. While I offered to assist on this for everyone, I have had basically 0 time over the past few weeks to look at any of the suggested changes and try to apply them.

I still want to try and do these fixes, but it's just a matter of getting time to look through it.

**Comment by UlisesFreitas** [comment]
No rush Silver I will try to review this template this week and try to solve the issues, I'll keep you updated here when I'm done with the changes. Thanks  to all for the help.

**Comment by Midhil457** [comment]
I would also like to help, if I have the time 😅

**Comment by 4ian** [comment]
A bit sad that we've fallen behind on this. I would say let's fix what we can in a few days then we can publish this as an example.

**Comment by Bouh** [comment]
I just opened the versions that have been shared in this thread.
There are two projects folder, I took the most recent one, the one from the first post that was updated on June 28.

A first review has been done by **Midhil**, none of the points listed have been fixed.
If a more updated version or one with the latest corrections that Midhil suggests is out there I will take another look.

**Comment by UlisesFreitas** [comment]
Please wait until I get those changes done and I re submit the new version

**Comment by Bouh** [comment]
> Please wait until I get those changes done and I re submit the new version

Ok thanks :)

I would add almost all points in the Midhil review come from[ the best practice guide](http://wiki.compilgames.net/doku.php/gdevelop5/community/guide-for-submitting-an-example), it wasn't up when the game has been submitted, but the rules were pretty similar.

**Comment by 4ian** [comment]
Thanks both :)

**Comment by UlisesFreitas** [comment]
Ok so I'm kinda make another approach 

https://games.gdevelop-app.com/game-73aebbdc-4b95-4444-9386-41c6beccc8c2/index.html
And here the new project 

https://github.com/UlisesFreitas/Shooty-GDevelop-example
Make sure the version you're going to review is 1.0.5

Well although it is not exactly the example that I wanted to share because not having nice assets the game is a bit bland, but hey there it goes.

I have changed the colors of the groups
I have renamed all variables and objects
removed the music
removed unused sounds
removed all third party assets there are only assets made by me
removed external layouts and all objects moved to Hud layout
Renamed the Layers
changed the behavior of enemies to chase the player 150 pixels away
Added some documentation in the comments.

If this example is not good, with your permission I am going to shoot myself in the face.
